File: epi_judge_java/epi/test_framework/TestUtils.java

package epi.test_framework;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Stream;

public class TestUtils {
  public static List<List<String>> splitTsvFile(Path tsvFile) {
    final String FIELD_DELIM = "\t";

    Stream<String> inputData = null;
    try {
      inputData = Files.lines(tsvFile);
    } catch (IOException e) {
      throw new RuntimeException("Test data file not found");
    }

    List<String> asList =
        inputData.collect(ArrayList::new, ArrayList::add, ArrayList::addAll);

    List<List<String>> result = new ArrayList<>();
    for (String line : asList) {
      result.add(List.of(line.split(FIELD_DELIM)));
    }
    return result;
  }

  public static String getDefaultTestDataDirPath() {
    final int MAX_SEARCH_DEPTH = 4;
    final String DIR_NAME = "test_data";

    Path path = Paths.get(".").toAbsolutePath();
    for (int i = 0; i < MAX_SEARCH_DEPTH; i++) {
      if (Files.isDirectory(path.resolve(DIR_NAME))) {
        return path.resolve(DIR_NAME).toString();
      }
      path = path.getParent();
      if (path == null) {
        break;
      }
    }

    throw new RuntimeException(
        "Can't find test data directory. Please start the program with \"--test_data_dir <path>\" command-line option");
  }

  public static String getFilePathInJudgeDir(String fileName) {
    return Paths.get(getDefaultTestDataDirPath())
        .getParent()
        .resolve(fileName)
        .toAbsolutePath()
        .toString();
  }

  /**
   * Serialized type name can contain multiple comments, enclosed into brackets.
   * This method removes all such comments.
   */
  public static String filterBracketComments(String s) {
    final String BRACKET_ENCLOSED_COMMENT = "(\\[[^\\]]*\\])";
    return s.replaceAll(BRACKET_ENCLOSED_COMMENT, "").replaceAll(" ", "");
  }

  /**
   * Check that result list has the same count of elements as reference.
   * TestFailure is thrown in case of mismatch.
   */
  public static <T> void assertAllValuesPresent(List<T> reference,
                                                List<T> result)
      throws TestFailure {
    Map<T, Integer> referenceSet = new HashMap<>();
    for (T x : reference) {
      referenceSet.put(x, referenceSet.getOrDefault(x, 0) + 1);
    }

    for (T x : result) {
      referenceSet.put(x, referenceSet.getOrDefault(x, 0) - 1);
    }

    List<T> excessItems = new ArrayList<>();
    List<T> missingItems = new ArrayList<>();

    referenceSet.forEach((x, count) -> {
      if (count < 0) {
        while (count++ < 0) {
          excessItems.add(x);
        }
      } else if (count > 0) {
        while (count-- > 0) {
          missingItems.add(x);
        }
      }
    });

    if (!excessItems.isEmpty() || !missingItems.isEmpty()) {
      TestFailure e =
          new TestFailure("Value set changed")
              .withProperty(TestFailure.PropertyName.RESULT, result);
      if (!excessItems.isEmpty()) {
        e.withProperty(TestFailure.PropertyName.EXCESS_ITEMS, excessItems);
      }
      if (!missingItems.isEmpty()) {
        e.withProperty(TestFailure.PropertyName.MISSING_ITEMS, missingItems);
      }
      throw e;
    }
  }

  /**
   * Non-exact float comparison
   */
  public static boolean floatComparison(Float f1, Float f2) {
    float eps = 1E-4f;
    float absEps = 1E-10f;
    return Math.abs(f1 - f2) <=
        Math.max(eps * Math.max(Math.abs(f1), Math.abs(f2)), absEps);
  }

  /**
   * Non-exact double comparison
   */
  public static boolean doubleComparison(Double d1, Double d2) {
    double eps = 1E-6;
    double absEps = 1E-20;
    return Math.abs(d1 - d2) <=
        Math.max(eps * Math.max(Math.abs(d1), Math.abs(d2)), absEps);
  }
}

File: epi_judge_python_solutions/test_framework/test_utils.py

import collections
import os
import re

from test_framework.test_failure import PropertyName, TestFailure


def split_tsv_file(tsv_file):
    ROW_DELIM = '\n'
    FIELD_DELIM = '\t'

    try:
        with open(tsv_file) as input_data:
            return [
                row.replace(ROW_DELIM, '').split(FIELD_DELIM)
                for row in input_data
            ]
    except OSError:
        raise RuntimeError('Test data file not found')


def get_default_test_data_dir_path():
    MAX_SEARCH_DEPTH = 4

    path = 'test_data'
    for _ in range(MAX_SEARCH_DEPTH):
        if os.path.isdir(path):
            return path
        path = os.path.join(os.path.pardir, path)

    raise RuntimeError(
        'Can\'t find test data directory. Please start the program with "--test_data_dir <path>" command-line option'
    )


def get_file_path_in_judge_dir(file_name):
    return os.path.join(get_default_test_data_dir_path(), os.path.pardir,
                        file_name)


def filter_bracket_comments(s):
    """
    Serialized type name can contain multiple comments, enclosed into brackets.
    This function removes all such comments.
    """
    bracket_enclosed_comment = r"(\[[^\]]*\])"
    return re.sub(bracket_enclosed_comment, '', s, 0).replace(' ', '')


def assert_all_values_present(reference, result):
    reference_set = collections.defaultdict(int)

    for x in reference:
        reference_set[x] += 1

    for x in result:
        reference_set[x] -= 1

    flatten = lambda l: [item for sublist in l for item in sublist]
    excess_items = flatten(
        [x] * -count for x, count in reference_set.items() if count < 0)
    missing_items = flatten(
        [x] * count for x, count in reference_set.items() if count > 0)

    if excess_items or missing_items:
        e = TestFailure('Value set changed')\
            .with_property(PropertyName.RESULT, result)
        if excess_items:
            e.with_property(PropertyName.EXCESS_ITEMS, excess_items)
        if missing_items:
            e.with_property(PropertyName.MISSING_ITEMS, missing_items)
        raise e


def completely_sorted(x):
    """Multi-dimensional container sort.

    If x is list of lists of ... of lists,
    then the argument is lexicographically sorted on all levels,
    starting from the bottom.
    """
    if isinstance(x, list):
        return sorted(completely_sorted(e) for e in x)
    else:
        return x


def unordered_compare(a, b):
    """Compares elements of 2 (multi-dimensional) list."""
    return completely_sorted(a) == completely_sorted(b)


def has_executor_hook(func):
    return hasattr(func, 'executor_hook') and func.executor_hook


def enable_executor_hook(func):
    func.executor_hook = True
    return func
